modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -387,17 +389,20 @@ private CompletableFuture<AsyncSqlCursor<List<Object>>> querySingle0(
                     return nodes.get(0);
                 })
                 .thenCompose(sqlNode -> {
+                    final boolean rwTx = dataModificationOp(sqlNode);
+
                     BaseQueryContext ctx = BaseQueryContext.builder()
                             .frameworkConfig(
                                     Frameworks.newConfigBuilder(FRAMEWORK_CONFIG)
                                             .defaultSchema(schema)
-                                            .traitDefs(Commons.LOCAL_TRAITS_SET)
+                                            .traitDefs(rwTx ? Commons.LOCAL_TRAITS_SET : Commons.DISTRIBUTED_TRAITS_SET)
                                             .build()
                             )
                             .logger(LOG)
                             .cancel(queryCancel)
                             .parameters(params)
                             .transaction(outerTx)
+                            .transactionTime(!rwTx ? (outerTx != null ? outerTx.readTimestamp() : new HybridClock().now()) : null)
```

Comment:
Will it be works in case outerTx exists and it's RW transaction?

Reply:
If this is rw tx we will set null here, is it a question ?

Reply:
I mean the following case:
1) we have an outer RW transaction
2) user do SELECT, so rwTx = false
3) you will set transaction time as not null